exercises/chapter5/test/no-peeking/Solutions.purs

```diff
@@ -0,0 +1,95 @@
+module Test.NoPeeking.Solutions where
+
+import Prelude
+
+import Data.Maybe (Maybe(Just, Nothing))
+import Data.Person (Person)
+import Data.Picture 
+  ( Bounds
+  , Picture
+  , Point(Point)
+  , Shape(Circle, Rectangle, Line, Text)
+  , origin
+  , bounds
+  , intersect
+  )
+import Data.Picture as DataP
+import Math as Math
+
+factorial :: Int -> Int
+factorial 0 = 1
+factorial n = n * factorial (n - 1)
+
+binomial :: Int -> Int -> Int
+binomial n 0 = 1
+binomial n m 
+  | n == m = 1
+  | otherwise = factorial n / (factorial m * (factorial (n - m)))
+
+{-
+Most general type for sameCity and livesInLA functions taking into account row polymorphism:
+
+sameCity 
+  :: forall r1 s1. { address :: { city :: String | s1 } | r1 } 
+  -> forall r2 s2. { address :: { city :: String | s2 } | r2 }
+  -> Boolean
+
+livesInLA 
+  :: forall r s. { address :: { city :: String | s } | r } 
+  -> Boolean
+-}
+
+sameCity :: Person -> Person -> Boolean
+sameCity { address: { city: c1 } } { address: { city: c2 } }
+  | c1 == c2 = true
+  | otherwise = false
+
+fromSingleton :: forall a. a -> Array a -> a
+fromSingleton a [] = a
+fromSingleton a [b] = b
+fromSingleton a _ = a
+
+circleAtOrigin :: Shape
+circleAtOrigin = Circle origin 10.0
+
+centerShape :: Shape -> Shape
+centerShape (Circle c r) = Circle origin r
+centerShape (Rectangle c w h) = Rectangle origin w h
+centerShape (Line (Point s) (Point e)) = 
+  (Line 
+    (Point { x: s.x - deltaX, y: s.y - deltaY })
+    (Point { x: e.x - deltaX, y: e.y - deltaY })
+  )
+  where
+  deltaX = (e.x + s.x) / 2.0
+  deltaY = (e.y + s.y) / 2.0
+centerShape (Text loc text) = Text origin text
+
+scaleShape :: Number -> Shape -> Shape
+scaleShape i (Circle c r) = Circle c (r * i)
+scaleShape i (Rectangle c w h) = Rectangle c (w * i) (h * i)
+scaleShape i (Line (Point s) (Point e)) = 
+  (Line 
+    (Point { x: s.x * i, y: s.y * i })
+    (Point { x: e.x * i, y: e.y * i })
+  )
+scaleShape i text = text
+
+doubleScaleAndCenter :: Shape -> Shape
+doubleScaleAndCenter = centerShape <<< scaleShape 2.0
+
+shapeText :: Shape -> Maybe String
+shapeText (Text _ text) = Just text
+shapeText _ = Nothing
+
+area :: Shape -> Number
+area (Circle _ r) = Math.pi * r * r
+area (Rectangle _ h w) = h * w
+area _ = 0.0
+
+data ShapeExt = Clipped Picture Point Number Number
```

Comment:
I'm not sure if this is the ideal solution. Based on this text from the exercise:

> Note that this makes Shape into a recursive data type.

It seems like the more appropriate solution would be to modify `Shape` in the `Picture.purs` file to look something like:
```
data Shape
  = Circle Point Number
  | Rectangle Point Number Number
  | Line Point Point
  | Text Point String
  | Clipped Picture Point Number Number
```
This would require modifying source outside of the `Solutions.purs` file, which seems to go against how these test exercises have been structured.

Reply:
Good call. Since the original exercises weren't written with unit testing in-mind, we'll have to use some workarounds like this. In many cases, it's tough to even make the exercises testable. I'd like to remove those if we can come up with substitute exercises.
But the compromise you came up with for this one looks good.